pkg/controller/data_upload_controller.go

```diff
@@ -350,29 +355,33 @@ func (r *DataUploadReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 }
 
-func (r *DataUploadReconciler) runCancelableDataUpload(ctx context.Context, asyncBR datapath.AsyncBR, du *velerov2alpha1api.DataUpload, res *exposer.ExposeResult, log logrus.FieldLogger) (reconcile.Result, error) {
-	log.Info("Run cancelable dataUpload")
+func (r *DataUploadReconciler) initCancelableDataPath(ctx context.Context, asyncBR datapath.AsyncBR, res *exposer.ExposeResult, log logrus.FieldLogger) error {
+	log.Info("Init cancelable dataUpload")
 
 	if err := asyncBR.Init(ctx, nil); err != nil {
-		return r.errorOut(ctx, du, err, "error to initialize asyncBR", log)
+		return errors.Wrap(err, "error initializing asyncBR")
 	}
 
 	log.Infof("async backup init for pod %s, volume %s", res.ByPod.HostingPod.Name, res.ByPod.VolumeName)
 
+	return nil
+}
+
+func (r *DataUploadReconciler) startCancelableDataPath(asyncBR datapath.AsyncBR, du *velerov2alpha1api.DataUpload, res *exposer.ExposeResult, log logrus.FieldLogger) error {
+	log.Info("Start cancelable dataUpload")
+
 	if err := asyncBR.StartBackup(datapath.AccessPoint{
 		ByPath: res.ByPod.VolumeName,
 	}, du.Spec.DataMoverConfig, nil); err != nil {
-		return r.errorOut(ctx, du, err, fmt.Sprintf("error starting async backup for pod %s, volume %s", res.ByPod.HostingPod, res.ByPod.VolumeName), log)
+		return errors.Wrapf(err, "error starting async backup for pod %s, volume %s", res.ByPod.HostingPod.Name, res.ByPod.VolumeName)
 	}
 
-	log.Infof("Async backup started for pod %s, volume %s", res.ByPod.HostingPod, res.ByPod.VolumeName)
-	return ctrl.Result{}, nil
+	log.Infof("Async backup started for pod %s, volume %s", res.ByPod.HostingPod.Name, res.ByPod.VolumeName)
+	return nil
 }
 
 func (r *DataUploadReconciler) OnDataUploadCompleted(ctx context.Context, namespace string, duName string, result datapath.Result) {
-	defer func() {
-		go r.closeDataPath(ctx, duName)
-	}()
+	defer r.dataPathMgr.RemoveAsyncBR(duName)
```

Comment:
It is under the context of callbacks which are called by the async backup/restore thread.
So calling `closeDataPath` which calls `asyncBR.Close` is not literally correct --- the async backup/restore thread is a part of the asyncBR instance so `asyncBR.Close` literally needs to first wait the thread and then close all the leftovers. However, this will cause a deadlock.
This is why we needed a go routine to call `closeDataPath` in the callbacks before this change. However, that is not a ideal solution.

Therefore, the logics are refactored:

- We create two functions for asyncBR, `close` and `Close`. The latter one calls the former one
- The async backup/restore thread calls `close` by the end itself
- So the callbacks never need to care about the closure of the asyncBR, they only need to take care of their own data
- `Close` is used by the caller to stop asyncBR outside of the async thread